test/integration.ts

```diff
@@ -0,0 +1,236 @@
+        })
+        const createResJson = await createRes.json()
+        log("Created product in core-api: %o", createResJson)
+        productId = createResJson.id
+        productIdBytes = "0x" + productId
+
+        // add stream to product
+        const streamIdEncoded = encodeURIComponent(streamId)
+        let status = 0
+        do {
+            // @ts-ignore-next-line private
+            const addRes = await ownerApiClient._post(`${STREAMR_API_URL}/products/${productId}/streams/${streamIdEncoded}`, {}, "PUT")
+            log("Add stream to product returned %s %s", addRes.status, addRes.statusText)
+            status = addRes.status
+        } while (status !== 204)
+
+        // createProduct(bytes32 id, string memory name, address beneficiary, uint pricePerSecond, Currency currency, uint minimumSubscriptionSeconds) public whenNotHalted {
+        const createTx = await market.createProduct(
+            productIdBytes,
+            "End-to-end tester",
+            watcherWallet.address,
+            1,  // pricePerSecond
+            0,  // DATA
+            1   // minimumSubscriptionSeconds
+        )
+        log("Creating product %s", productId)
+        const createTr = await createTx.wait()
+
+        // set it deployed (simulate watcher)
+        // @ts-ignore-next-line private
+        const deployRes = await trustedApiClient._post(`${STREAMR_API_URL}/products/${productId}/setDeployed`, {
+            blockNumber: createTr.blockNumber,
+            blockIndex: createTr.transactionIndex,
+            beneficiaryAddress: watcherWallet.address,
+            description: "Testing!",
+            name: "Test",
+            ownerAddress: watcherWallet.address,
+            minimumSubscriptionInSeconds: 1,
+            pricePerSecond: 1,
+            priceCurrency: "DATA",
+            category: "1"
+        })
+        log("Set deployed in core-api: %o", await deployRes.json())
+    })
+
+    let watcherProcess: ChildProcess | undefined
+    beforeEach(async function () {
+        this.timeout(10000)
+        if (watcherProcess) { return }
+        log(Object.keys(watcherEnv).map((k) => k + "=" + watcherEnv[k] + " ").join("") + watcherExecutable)
+        log(process.execPath)
+        watcherProcess = spawn(process.execPath, [watcherExecutable], { env: watcherEnv })
+        if (!watcherProcess.stdout || !watcherProcess.stderr) { throw new Error("No stdout/stderr in spawned process") }
+        watcherProcess.stdout.on("data", (data) => { log(`(server stdout) ${data.toString().trim()}`) })
+        watcherProcess.stderr.on("data", (data) => { log(`(server stderr) ${data.toString().trim()}`) })
+        watcherProcess.on("close", (code) => { log(`server exited with code ${code}`) })
+        watcherProcess.on("error", (err) => { log(`server ERROR: ${err}`) })
+
+        // wait until server is started, also get url from output
+        await untilStreamMatches(watcherProcess.stdout, /Playback done. Starting watcher.../)
```

Comment:
Reading process stdout is probably fine, but it's also brittle. Would it be possible to have some event functions in the watcher that you register to listen and do a callback function. Ie. you get notified by the watcher that this thing just happened and your callback function gets called.

Reply:
not sure how that works over processes...

The point of starting a separate process is, I guess, to test it as it would be running on real server: a separate process. Of course creating a watcher etc. in the test would make it more poke-able but also less realistic.

Starting a separate process is a straightforward way to test running the js file as a script. I guess you can run it in the same process too (`require` does something like that I guess?).

Reply:
That's just an idea. I would still consider it an integration test if it would be something like (pseudo):
```
const watcher = new Watcher(function onServerStarted() {
    runIntegrationTest()
})
```

Reply:
I'd call it an integration test even if it disregards things like command line arguments or doesn't read environment variables in `main` function.

Reply:
sure, but spawning that process isn't a big deal IMO. Not as pretty as something like perl maybe but still okay.